x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/history/ILMHistoryStore.java

```diff
@@ -44,98 +45,119 @@
 
 /**
  * The {@link ILMHistoryStore} handles indexing {@link ILMHistoryItem} documents into the
- * appropriate index. It sets up a {@link BulkProcessor} for indexing in bulk, and handles creation
+ * appropriate index. It sets up a {@link BulkProcessor2} for indexing in bulk, and handles creation
  * of the index/alias as needed for ILM policies.
  */
 public class ILMHistoryStore implements Closeable {
     private static final Logger logger = LogManager.getLogger(ILMHistoryStore.class);
 
+    public static final Setting<Integer> MAX_BULK_REQUEST_QUEUE_SIZE_SETTING = Setting.intSetting(
+        "es.indices.lifecycle.history.bulk.request.queue.size",
+        100,
+        1,
+        Setting.Property.NodeScope
+    );
```

Comment:
I think it would be preferable to bound the total _size_ of all incomplete requests rather than their number. We kind of don't mind if there are 1000+ requests in progress as long as they're all small, and conversely 100 max-sized requests is still 500MiB of heap which is an awful lot of memory to dedicate to this component.

An overall bound on resources used by the whole system would also avoid having to think about bounds at each level as we do today.

Reply:
Yeah we had talked about this being a problem with BulkProcessor, and settled on the solution being that each user would decide how big its queue would be (so here the ILM history store is defaulting to 100 because its requests will be relatively small and few). Sounds like you're suggesting instead that instead of one BulkProcessor queue per user we have one for the whole system, and bound it by size in bytes rather than count?

Reply:
I'm ok with the limit being per instance, I just think the limit should be expressed in approximate-bytes-of-heap rather than items-in-the-queue.